File: leetcode/0599-minimum-index-sum-of-two-lists/solution.cpp
class Solution {
public:
    /**
     * RUNTIME: 259 ms (35.90%)
     * MEMORY: 45.9 MB (5.25%)
    */
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        int least = list1.size() + list2.size();
        unordered_map<string, int> list_dict1, list_dict2;
        unordered_map<int, vector<string>> least_dict;
        if (list1.size() > list2.size()) {
            vector<string> tmp;
            tmp = list1;
            list1 = list2;
            list2 = tmp;
        }
        for (int i=0; i < list2.size(); i++) {
            list_dict2[list2[i]] = i;
        }
        for (int i=0; i < list1.size(); i++) {
            if (list_dict2.find(list1[i]) != list_dict2.end()) {
                list_dict1[list1[i]] = i;
            }
        }
        for (auto i : list_dict1) {
            string str = i.first;
            if (list_dict2.find(str) != list_dict2.end()) {
                int min_sum = list_dict1[str] + list_dict2[str];
                if (least > min_sum) {
                    least_dict.clear();
                    vector<string> tmp;
                    tmp.push_back(str);
                    least_dict[min_sum] = tmp;
                    least = min_sum;
                }
                else if (least == min_sum) {
                    least_dict[min_sum].push_back(str);
                }
            }
        }
        return least_dict[least];
    }
};

File: leetcode/0599-minimum-index-sum-of-two-lists/solution.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        """
        RUNTIME: 271 ms (79.87%)
        MEMORY: 14.5 MB (40.48%)
        """
        least = len(list1) + len(list2)
        list_dict1, list_dict2 = {}, {}
        if len(list1) > len(list2):  # for tweaking running time
            list1, list2 = list2, list1
        for i, string in enumerate(list2):
            list_dict2[string] = i
        for i, string in enumerate(list1):
            if string in list_dict2:  # for tweaking running time
                list_dict1[string] = i

        for string in list_dict1:
            if string in list_dict2:
                min_sum = list_dict1[string] + list_dict2[string]
                if least > min_sum:
                    least_dict = {min_sum: [string]}
                    least = min_sum
                elif least == min_sum:
                    least_dict[min_sum].append(string)
        return least_dict[least]
